### PR TITLE
CSHARP-3919: Drivers should retry operations if connection handshake fails

### PR DESCRIPTION
**Pull request by BorisDog**
[EG](https://spruce.mongodb.com/version/62211c2bd6d80a1ce9d9be96/tasks)

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs [hunk 1]
```diff
@@ -160,6 +160,9 @@ public static bool IsRetryableReadException(Exception exception)
             return false;
         }
 
+        public static bool IsRetryableHandshakeException(Exception exception) =>
+            IsNetworkException(exception) || IsRetryableReadException(exception);
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs:164]: `IsNetworkException` is included in `IsRetryableReadException`

File: src/MongoDB.Driver.Core/Core/Operations/RetryableReadContext.cs [hunk 2]
```diff
@@ -183,40 +183,25 @@ private void Initialize(CancellationToken cancellationToken)
...
-                }
+                ReplaceChannelSource(_binding.GetReadChannelSource(cancellationToken));
+                ReplaceChannel(_channelSource.GetChannel(cancellationToken));
             }
         }
 
         private async Task InitializeAsync(CancellationToken cancellationToken)
         {
             _channelSource = await _binding.GetReadChannelSourceAsync(cancellationToken).ConfigureAwait(false);
-            var serverDescription = _channelSource.ServerDescription;
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/RetryableReadContext.cs:203]: same change is needed for sync version

File: src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs (hunk 1 above)
Comment (BorisDog) [reply src/MongoDB.Driver.Core/Core/Operations/RetryabilityHelper.cs]: Good catch. Done.

File: tests/MongoDB.Driver.Tests/Specifications/retryable-reads/RetryableReadsUnifiedTestRunner.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
+﻿/* Copyright 2021-present MongoDB Inc.
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply tests/MongoDB.Driver.Tests/Specifications/retryable-reads/RetryableReadsUnifiedTestRunner.cs]: Done.

File: tests/MongoDB.Driver.Tests/Specifications/retryable-writes/RetryableWritesUnifiedTestRunner.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            {
+                runner.Run(testCase);
+            }
+        }
+
+        // nested types
+        public class TestCaseFactory : JsonDrivenTestCaseFactory
+        {
+            // protected properties
+            // the path is "retryable-writes-reads" but the namespace is "retryable_writes"
```
[the remark this answers is not in the document]
Comment (BorisDog) [reply tests/MongoDB.Driver.Tests/Specifications/retryable-writes/RetryableWritesUnifiedTestRunner.cs]: Removed.

**Review by DmitryLukyanov — APPROVED** [review]
LGTM

**Pull request merged**